### ISSUE TITLE
Why do variables.env and configuratin.py us different dns names for the same services?

### ISSUE THREAD

**Issue by freemansoft** [NVIDIA/nim-anywhere#27]
What are the DNS names for the redis and milvus server?  They are referred to as `localhost` in some places and as `redis` and `mivlus` in others.  They are pingable from `traefic` as `redis` and `milvus`

#### Redis
1. **README.md** redis://localhost:6379/0
1. **variables.env** redis://redis:6379/0
1. **chain_server/configuration.py**  redis://localhost:6379/0

#### Milvus
1. **README.md** http://localhost:19530
2. **variables.env** http://milvus:19530
3. **chain_server/configuration.py** http://localhost:19530

**Comment by brainwavecoder9** [comment]
The host names are `redis` and `milvus` but it can be more complex because the containerized aspect introduces additional scopes. The short version is that it depends on where you are referencing them from. Generally you can think of this as 3 scopes:  

1. **The host where you are running AI Workbench.** In many cases you won't have access to this, or at least, it's less relevant because your AI Workbench environment will be entirely inside of it's own container.
3. **The AI Workbench container.** This is what you see being built, and where everything happens when you run your environment. In most cases, this is is the only reference you will even be aware of.
4. **Other containers / networks / bindings / etc.** In some cases, such as this NIM Anywhere project, an approach is used where multiple containers are at play. This is atypical and less common, but is possible to encounter. There are a number of ways this can be accomplished so specifics will depend on each unique deployment, but in this case redis/milivus are each their own containers (and you are able to access them from within the AI Workbench container). Note that per documentation the current approach used for NIM Anywhere is expected to change. I'm not tapped into any details, but I would assume they will introduce some type of orchestration approach which will allow your AI Workbench environment to spin up multiple containers instead of just the one we have today.

So... given all of that, the reference to a given hostname or "localhost" could actually mean: The actual host, The AI Workbench Container, one of the other containers, and a localhost port could actually be bound between the host-container or another container. The short version is that there are a lot of ways this can be accomplished, so each setup could have it's own unique approach.

Since there's not a single common way of doing this, and because there are so many potential configurations including exposing and mapping a port outside of a container (either on the host, a container, a container network, or other) you need to figure out where you are and look at what you are referencing to understand what you are actually looking at.

If you want to go down the rabbit hole, you can find more info in the [Docker Networking docs](https://docs.docker.com/engine/network/), but that doesn't even get into more abstract configurations like Docker-in-Docker or Docker-out-of-Docker (which is what NIM Anywhere uses); these and other approaches add additional potential layers of complexity.

**Comment by freemansoft** [comment]
That is an awesome explanation. It still confuses me about the configuration _in this project_  

How are the different files intended to be consumed with their different values? It isn't obvious what processes use which file under which conditions.

* `variables.env`
    * Config for `milvus` defaults to `milvus:`.  Doesn't seem to be referenced.
    * Config for `redis` defaults to `redis:`.  Doesn't seem to be referenced.
* `frontend/configuration.py`,
    * Not part of this
* `code/config.yaml` 
    * Config for `llm-0:` defaults to container name
    * Config for `embedding` defaults to the container name
    * Config for `reranking` defaults to the container name 
    * It uses the _container_ name for locally running containers and **not** `localhost`
* `chain_server/config.yaml` 
    * Config for `milvus` defaults to `localhost`
    * Config for `redis` defaults to `localhost`
    * Config for  `llm` defaults to NVIDIA endpoints or the container name if local.
    * Config for `re-ranker` defaults to NVIDIA endpoints or the container name if local.
    * Cnofig for `embedding` defaults to NVIDIA endpoints or the container name if local.

Healthchecks point at `localhost`.  This is because they test from inside the target container.

The NIM-ANywhere project is a multi-container project. The AI Workbench container sees the Milvus and Redis containers as their container names because they are part of the same docker instance.  There are 6 services in this project that all run in their own containers or are deployed remotely.  Why are they not in variables.env? 

* milvus.  Runs locally. Avaialble in-cluster `milvus`. 
    *  👍 Connect from `project-nim-anywhere` to `milvus:6379`
    *  👎 Cannot connect from `project-nim-anywhere` to `localhost:6389`
* redis.  Runs locally.  Available in-cluster under `redis`.   
    *  👍 Connect from `project-nim-anywhere` to `redis:19530`
    *  👎 Connect from `project-nim-anywhere` to `localhost:19530`

**Comment by brainwavecoder9** [comment]
Reading a little between the lines I think you're ultimately asking the question "Why not just add the hostnames in one place like variables.env, instead of the references to localhost or other places, because project-nim-anywhere container can't access them as localhost anyways?"

Defining them in variables.env is certainly a more robust approach that would provide better flexibility and maintainability for a larger or more complex system, but for this interim stage, simplicity wins. It's easy to hard code "milvus"... maybe even easier than managing a global config and referencing something like "$MILVUS_HOSTNAME"

As for wondering which file is used under which condition, I think what's in place is misleading. Inside _any_ container, localhost refers only to services within that container (not other containers) so I don't think references to localhost are actually used. localhost would make sense with chain-server deployed on a traditional server, but not in this setup. Any references to localhost are likely just unused remnants of default configuration for a non-containerized deployment. 

Since it's working the team probably didn't feel compelled to bother cleaning up the extra bits... even if what's in place isn't ideal for any particular reason. As for why any self respecting team would choose to move forward in such a state of disarray, I can't help but think of a phrase that an old developer friend used to say, and that I fully agree with: "Premature optimization is the root of all evil."

**tl;dr - Don't look a gift horse in the mouth. References to localhost probably aren't used. Since container names are simple, and [all containers are networked together](https://github.com/NVIDIA/nim-anywhere/blob/409f868a0f4dd9cf06283ef7bb1c357462ccd101/apps/functions#L36C1), it's easy just to hard code the host/container names where needed and call it a day.**

**Comment by brainwavecoder9** [comment]
PS - We do see an example in the [README ](https://github.com/NVIDIA/nim-anywhere/blob/409f868a0f4dd9cf06283ef7bb1c357462ccd101/README.md?plain=1#L578) where `localhost` is identified, but the override was defined in [variables.env](https://github.com/NVIDIA/nim-anywhere/blob/409f868a0f4dd9cf06283ef7bb1c357462ccd101/variables.env#L7).

"Values specified as environment variables will take precedence over all values from files."
